File: 2017/Advent25.java
import java.util.*;
import java.util.stream.*;

public class Advent25 extends Advent {

  int position = 0;
  Map<Integer, Integer> tape = new TreeMap<>();
  State state = State.A;
  int steps = 12794428;


  public Advent25() {
    super(25);
  }

  @Override
  protected void readInput() {

  }

  @Override
  protected void parseInput() {

  }

  @Override
  protected String part1() {
    for (int i = 0; i < steps; i++)
      eval();
    return "" + checksum();
  }

  @Override
  protected String part2() {
    return "Not complete";
  }

  void eval() {
    int value = read();
    switch (state) {
      case A:
        if (value == 0) {
          write(1);
          move(Direction.RIGHT);
          state = State.B;
        } else {
          write(0);
          move(Direction.LEFT);
          state = State.F;
        }
        break;
      case B:
        if (value == 0) {
          write(0);
          move(Direction.RIGHT);
          state = State.C;
        } else {
          write(0);
          move(Direction.RIGHT);
          state = State.D;
        }
        break;
      case C:
        if (value == 0) {
          write(1);
          move(Direction.LEFT);
          state = State.D;
        } else {
          write(1);
          move(Direction.RIGHT);
          state = State.E;
        }
        break;
      case D:
        if (value == 0) {
          write(0);
          move(Direction.LEFT);
          state = State.E;
        } else {
          write(0);
          move(Direction.LEFT);
          state = State.D;
        }
        break;
      case E:
        if (value == 0) {
          write(0);
          move(Direction.RIGHT);
          state = State.A;
        } else {
          write(1);
          move(Direction.RIGHT);
          state = State.C;
        }
        break;
      case F:
        if (value == 0) {
          write(1);
          move(Direction.LEFT);
          state = State.A;
        } else {
          write(1);
          move(Direction.RIGHT);
          state = State.A;
        }
        break;
    }
  }

  int read() {
    return tape.getOrDefault(position, 0);
  }

  void write(int value) {
    tape.put(position, value);
  }

  void move(Direction direction) {
    switch (direction) {
      case RIGHT:
        position++;
        break;
      case LEFT:
        position--;
        break;
    }
  }

  int checksum() {
    return tape.values().stream().mapToInt(e -> e.intValue()).sum();
  }

  enum State {
    A,
    B,
    C,
    D,
    E,
    F;
  }

  enum Direction {
    LEFT,
    RIGHT;
  }
}

File: 2017/python/advent_25.py
from common.advent_lib import *
from collections import Counter, defaultdict, deque
import collections as coll
import functools as func
import itertools as it
import operator as oper
import re
import more_itertools as it2
import numpy as np
import sympy as sym

chunks = full_input(25)[:-1].split("\n\n")

init=chunks[0].split("\n")
start_state=init[0].split()[-1][:-1]
steps=int(init[1].split()[-2])

states=dict()
for chunk in chunks[1:]:
    lines=[line[:-1].split()[-1] for line in chunk.split("\n")]
    state=lines[0]
    when_0 = tuple(lines[i] for i in [2,3,4])
    when_1 = tuple(lines[i] for i in [6,7,8])
    states[state]=(when_0,when_1)

tape=defaultdict(int)
pos=0
state=start_state
for _ in range(steps):
    value=tape[pos]
    write,direction,next_state=states[state][value]
    tape[pos]=int(write)
    pos+=1 if direction=="right" else -1
    state=next_state

part1 = sum(tape.values())
print("Part 1:", part1)

part2 = "All done!"
print("Part 2:", part2)
